test/NonFungibleToken.test.js

```diff
@@ -0,0 +1,898 @@
+
+                logs.length.should.be.equal(1);
+                logs[0].event.should.be.eq('Approval');
+                logs[0].args._owner.should.be.equal(sender);
+                logs[0].args._approved.should.be.equal(to);
+                logs[0].args._tokenId.should.be.bignumber.equal(tokenId);
+              });
+            });
+
+            describe('when the given token ID was approved for another account', function () {
+              beforeEach(async function () {
+                await token.approve(accounts[2], tokenId, { from: sender });
+              });
+
+              it('changes the approval to the given address', async function () {
+                await token.approve(to, tokenId, { from: sender });
+
+                const approvedAccount = await token.approvedFor(tokenId);
+                approvedAccount.should.be.equal(to);
+              });
+
+              it('emits an approval event', async function () {
+                const { logs } = await token.approve(to, tokenId, { from: sender });
+
+                logs.length.should.be.equal(1);
+                logs[0].event.should.be.eq('Approval');
+                logs[0].args._owner.should.be.equal(sender);
+                logs[0].args._approved.should.be.equal(to);
+                logs[0].args._tokenId.should.be.bignumber.equal(tokenId);
+              });
+            });
+          });
+
+          describe('when the address that receives the approval is the owner', function () {
+            const to = _creator;
+
+            describe('when there was no approval for the given token ID before', function () {
+              it('reverts', async function () {
+                await assertRevert(token.approve(to, tokenId, { from: sender }));
+              });
+            });
+
+            describe('when the given token ID was approved for another account', function () {
+              beforeEach(async function () {
+                await token.approve(accounts[2], tokenId, { from: sender });
+              });
+
+              it('reverts', async function () {
+                await assertRevert(token.approve(to, tokenId, { from: sender }));
+              });
+            });
+          });
+        });
+      });
+
+      describe('when the sender does not own the given token ID', function () {
+        const sender = accounts[1];
+
+        it('reverts', async function () {
+          await assertRevert(token.approve(accounts[2], tokenId, { from: sender }));
+        });
+      });
+    });
+
+    describe('when the given token ID was not tracked by the contract before', function () {
+      const tokenId = _unknownTokenId;
+
+      it('reverts', async function () {
+        await assertRevert(token.approve(accounts[1], tokenId, { from: _creator }));
+      });
+    });
+  });
+
+  describe('approveAll', function () {
+    const sender = _creator;
+
+    describe('when the address willing to approve is the 0 address', function () {
```

Comment:
is it worth including this test case? the code doesn't react differently based on `_to` being the 0-address or not

(up to you)